plotly/dashboard_objs/dashboard_objs.py

```diff
@@ -0,0 +1,302 @@
+"""
+dashboard_objs
+==========
+
+A module which is used to create dashboard objects, manipulate them and then
+upload them.
+
+"""
+
+import copy
+import json
+import requests
+import pprint
+import webbrowser
+
+import plotly
+
+from plotly import exceptions
+from plotly.utils import node_generator
+from plotly.api.v2.utils import build_url
+
+username = plotly.tools.get_credentials_file()['username']
+api_key = plotly.tools.get_credentials_file()['api_key']
+headers = {'Plotly-Client-Platform': 'nteract'}
+
+
+class EmptyBox(dict):
+    def __init__(self):
+        self['type'] = 'box'
+        self['boxType'] = 'empty'
+
+
+class Box(dict):
+    def __init__(self, fileId='', shareKey=None, title=''):
+        self['type'] = 'box'
+        self['boxType'] = 'plot'
+        self['fileId'] = fileId
+        self['shareKey'] = shareKey
+        self['title'] = title
+
+
+class Container(dict):
+    def __init__(self, box_1=EmptyBox(), box_2=EmptyBox(), size=400,
+                 sizeUnit='px', direction='vertical'):
+        self['type'] = 'split'
+        self['size'] = size
+        self['sizeUnit'] = sizeUnit
+        self['direction'] = direction
+        self['first'] = box_1
+        self['second'] = box_2
+
+
+class Dashboard(dict):
+    def __init__(self, dashboard_json=None, backgroundColor='#FFFFFF',
+                 boxBackgroundColor='#ffffff', boxBorderColor='#d8d8d8',
+                 boxHeaderBackgroundColor='#f8f8f8', foregroundColor='#333333',
+                 headerBackgroundColor='#2E3A46', headerForegroundColor='#FFFFFF',
+                 links=[], logoUrl='', title='Untitled Dashboard'):
```

Comment:
The more I look at all these, the more I wonder whether we need non-native objects at all? We're just using these as dicts and the user isn't meant to interact with anything here...

What do you think of just having a top-level *interface* to an underlying dict? I.e., the `Dashboard` is just an interface to make it easier to manipulate a dashboard dict?

Initialization would look something like this:

```
class Dashboard(object):

    def __init__(self, content=None):
        if content is None:
            content = {}
        if not isinstance(content, dict):
            raise TypeError('Content must be a dict.')

        # Ensure top-level attributes exist.
        content['version'] = 2
        if content.get('settings') is None:
            content['settings'] = {}
        if content.get('layout') is None:
            content['layout'] = {}

        # Defer initialization of layout to general method.
        self._initialize_split(content['layout'])

        self.content = content

    def _initialize_box(self, content):
        # Ensure top-level attributes exist.
        content['type'] = 'box'

        # yadda, yadda, more logic to decide what *kind*
        # of box this is and make sure to initialize it right.

    def _initialize_split(self, content):
        # Ensure top-level attributes exist.
        content['type'] = 'split'
        if content.get('first') is None:
            content['first'] = {}
        if content.get('second') is None:
            content['second'] = {}

        # yadda, yadda, more logic to decide what else we
        # need to properly initialize the top-level attributes of
        # the split

        # Defer initialization to general methods.
        if content['first'].get('type') == 'split':
            self._initialize_split(content['first']):
        else:
            self._initialize_box(content['first'])

        if content['second'].get('type') == 'split':
            self._initialize_split(content['second']):
        else:
            self._initialize_box(content['second'])
```

Pros:

* No need to inherit from `dict`, which means we have a handy separation-of-concerns. I.e., users create a `Dashboard` instance so that they can more-easily manipulate the `content`, which is a simple `dict` (no frills at all).
* No confusing call signatures. The less-specific we are, the better here. We want as many details as possible to be handled by the backend API *not* by the api wrapper here.
* Only one new object for users to grok.